retrieval/discovery/ec2.go

```diff
@@ -110,12 +115,25 @@ func (ed *EC2Discovery) refresh() (*config.TargetGroup, error) {
 				labels := model.LabelSet{
 					ec2LabelInstanceID: model.LabelValue(*inst.InstanceId),
 				}
+				labels[ec2LabelPrivateIP] = model.LabelValue(*inst.PrivateIpAddress)
+				addr := fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port)
+
 				if inst.PublicIpAddress != nil {
 					labels[ec2LabelPublicIP] = model.LabelValue(*inst.PublicIpAddress)
+					labels[ec2LabelPublicDns] = model.LabelValue(*inst.PublicDnsName)
+					// replace address with public dns name if present
+					addr = fmt.Sprintf("%s:%d", *inst.PublicDnsName, ed.port)
```

Comment:
> Still I understand your concern about the private ip convention, would a flag in ec2sdconfig here, for instance "use_hostname: true", be ok?

Our convention is that anything like that should be handled by relabelling, as there's just too many possible ways a user could want to configure things for us to offer options for all of them. This is something that may be okay in your particular setup, but isn't necessarily correct everywhere (e.g. if I only have a handful of hosts with public IPs in a VPC doing NAT or similar, then this behaviour is incorrect). In general we'd encourage a setup where the prometheus would be within the VPC, using the private IP.